diff --git a/scripts/pylib/twister/twisterlib/environment.py b/scripts/pylib/twister/twisterlib/environment.py
index 78e3f03720b..ccfde532f83 100644
--- a/scripts/pylib/twister/twisterlib/environment.py
+++ b/scripts/pylib/twister/twisterlib/environment.py
@@ -680,6 +680,11 @@ structure in the main Zephyr tree: boards/<vendor>/<board_name>/""")
              " risk. It is recommended to only use this option for local "
              " development and when testing localized change in a subsystem.")
 
+    parser.add_argument("--failure-script",
+                        help="""specify a failures script. This will be executed
+                        on test failure.
+                        """)
+
     parser.add_argument(
         '--detailed-test-id', action='store_true',
         help="Compose each test Suite name from its configuration path (relative to root) and "
diff --git a/scripts/pylib/twister/twisterlib/handlers.py b/scripts/pylib/twister/twisterlib/handlers.py
index 695888cc695..e59c1a6d756 100755
--- a/scripts/pylib/twister/twisterlib/handlers.py
+++ b/scripts/pylib/twister/twisterlib/handlers.py
@@ -543,6 +543,12 @@ class DeviceHandler(Handler):
     def make_dut_available(self, dut):
         if self.instance.status in [TwisterStatus.ERROR, TwisterStatus.FAIL]:
             dut.failures_increment()
+            if dut.failure_script:
+                logger.debug(f"Running failure script: {dut.failure_script}")
+                timeout = 30
+                if dut.script_param:
+                    timeout = dut.script_param.get("failure_script_timeout", timeout)
+                self.run_custom_script(re.split('[, ]', dut.failure_script), timeout=timeout)
         logger.debug(f"Release DUT:{dut.platform}, Id:{dut.id}, "
                      f"counter:{dut.counter}, failures:{dut.failures}")
         # get all DUTs with the same id
diff --git a/scripts/pylib/twister/twisterlib/hardwaremap.py b/scripts/pylib/twister/twisterlib/hardwaremap.py
index 919f9bfeb4c..9f1a8ddad70 100644
--- a/scripts/pylib/twister/twisterlib/hardwaremap.py
+++ b/scripts/pylib/twister/twisterlib/hardwaremap.py
@@ -49,7 +49,8 @@ class DUT:
                  runner=None,
                  flash_timeout=60,
                  flash_with_test=False,
-                 flash_before=False):
+                 flash_before=False,
+                 failure_script=None):
 
         self.serial = serial
         self.baud = serial_baud or 115200
@@ -70,6 +71,7 @@ class DUT:
         self.post_script = post_script
         self.pre_script = pre_script
         self.script_param = script_param
+        self.failure_script = failure_script
         self.probe_id = None
         self.notes = None
         self.lock = Lock()
@@ -199,6 +201,7 @@ class HardwareMap:
                 self.add_device(self.options.device_serial,
                                 self.options.platform[0],
                                 self.options.pre_script,
+                                self.options.failure_script,
                                 False,
                                 baud=self.options.device_serial_baud,
                                 flash_timeout=self.options.device_flash_timeout,
@@ -210,6 +213,7 @@ class HardwareMap:
                 self.add_device(self.options.device_serial_pty,
                                 self.options.platform[0],
                                 self.options.pre_script,
+                                self.options.failure_script,
                                 True,
                                 flash_timeout=self.options.device_flash_timeout,
                                 flash_with_test=self.options.device_flash_with_test,
@@ -239,6 +243,7 @@ class HardwareMap:
         serial,
         platform,
         pre_script,
+        failure_script,
         is_pty,
         baud=None,
         flash_timeout=60,
@@ -249,6 +254,7 @@ class HardwareMap:
             platform=platform,
             connected=True,
             pre_script=pre_script,
+            failure_script=failure_script,
             serial_baud=baud,
             flash_timeout=flash_timeout,
             flash_with_test=flash_with_test,
@@ -292,6 +298,7 @@ class HardwareMap:
             product = dut.get('product')
             fixtures = dut.get('fixtures', [])
             connected = dut.get('connected') and ((serial or serial_pty) is not None)
+            failure_script = dut.get('failure_script')
             if not connected:
                 continue
             for plat in platforms:
@@ -310,7 +317,8 @@ class HardwareMap:
                               post_flash_script=post_flash_script,
                               script_param=script_param,
                               flash_timeout=flash_timeout,
-                              flash_with_test=flash_with_test)
+                              flash_with_test=flash_with_test,
+                              failure_script=failure_script)
                 new_dut.fixtures = fixtures
                 new_dut.counter = 0
                 self.duts.append(new_dut)
diff --git a/scripts/schemas/twister/hwmap-schema.yaml b/scripts/schemas/twister/hwmap-schema.yaml
index 142d4a1969b..f369a84f3f3 100644
--- a/scripts/schemas/twister/hwmap-schema.yaml
+++ b/scripts/schemas/twister/hwmap-schema.yaml
@@ -50,6 +50,9 @@ sequence:
       "pre_script":
         type: str
         required: false
+      "failure_script":
+        type: str
+        required: false
       "fixtures":
         type: seq
         required: false
@@ -77,3 +80,6 @@ sequence:
           "post_flash_timeout":
             type: int
             required: false
+          "failure_script_timeout":
+            type: int
+            required: false
diff --git a/scripts/tests/twister/test_handlers.py b/scripts/tests/twister/test_handlers.py
index 49139da52d2..ef00cbfe841 100644
--- a/scripts/tests/twister/test_handlers.py
+++ b/scripts/tests/twister/test_handlers.py
@@ -1441,6 +1441,7 @@ def test_devicehandler_handle(
         pre_script='dummy pre script',
         post_script='dummy post script',
         post_flash_script='dummy post flash script',
+        failure_script='dummy failure script',
         flash_timeout=60,
         flash_with_test=True
     )
@@ -1498,7 +1499,8 @@ def test_devicehandler_handle(
     handler.run_custom_script.assert_has_calls([
         mock.call('dummy pre script', mock.ANY),
         mock.call('dummy post flash script', mock.ANY),
-        mock.call('dummy post script', mock.ANY)
+        mock.call('dummy post script', mock.ANY),
+        mock.call('dummy failure script', mock.ANY)
     ])
 
     if expected_reason:
diff --git a/scripts/tests/twister/test_hardwaremap.py b/scripts/tests/twister/test_hardwaremap.py
index 5d0ea798ed6..8b749adc0eb 100644
--- a/scripts/tests/twister/test_hardwaremap.py
+++ b/scripts/tests/twister/test_hardwaremap.py
@@ -53,6 +53,7 @@ TESTDATA_1 = [
             'pre_script': 'dummy pre script',
             'post_script': 'dummy post script',
             'post_flash_script': 'dummy post flash script',
+            'failure_script': 'dummy failure script',
             'runner': 'dummy runner',
             'flash_timeout': 30,
             'flash_with_test': True,
@@ -60,6 +61,7 @@ TESTDATA_1 = [
                 'pre_script_timeout' : 30,
                 'post_flash_timeout' : 30,
                 'post_script_timeout' : 30,
+                'failure_script_timeout' : 30,
                 }
         },
         {
@@ -75,6 +77,7 @@ TESTDATA_1 = [
             'pre_script': 'dummy pre script',
             'post_script': 'dummy post script',
             'post_flash_script': 'dummy post flash script',
+            'failure_script': 'dummy failure script',
             'runner': 'dummy runner',
             'flash_timeout': 30,
             'flash_with_test': True,
@@ -82,6 +85,7 @@ TESTDATA_1 = [
                 'pre_script_timeout' : 30,
                 'post_flash_timeout' : 30,
                 'post_script_timeout' : 30,
+                'failure_script_timeout' : 30,
                 }
         },
         '<dummy platform (dummy product) on dummy serial>'
