diff --git a/tests/drivers/dma/chan_blen_transfer/Kconfig b/tests/drivers/dma/chan_blen_transfer/Kconfig
index 8be4efa636a..3a739c8ff0c 100644
--- a/tests/drivers/dma/chan_blen_transfer/Kconfig
+++ b/tests/drivers/dma/chan_blen_transfer/Kconfig
@@ -25,3 +25,7 @@ config DMA_LOOP_TRANSFER_NUMBER_OF_DMAS
 config DMA_TRANSFER_BURST16
 	bool "Enable loop transfers of 16-beat bursts"
 	default y
+
+config DMA_BLEN_ALIGNMENT
+	int "TX and RX data alignment in bytes"
+	default 32
diff --git a/tests/drivers/dma/chan_blen_transfer/src/test_buffers.c b/tests/drivers/dma/chan_blen_transfer/src/test_buffers.c
index 3ec893f15c3..36c1fb81409 100644
--- a/tests/drivers/dma/chan_blen_transfer/src/test_buffers.c
+++ b/tests/drivers/dma/chan_blen_transfer/src/test_buffers.c
@@ -8,5 +8,6 @@
 
 #include "test_buffers.h"
 
-__aligned(32) char tx_data[TEST_BUF_SIZE] = "It is harder to be kind than to be wise........";
-__aligned(32) char rx_data[TEST_BUF_SIZE] = { 0 };
+__aligned(CONFIG_DMA_BLEN_ALIGNMENT) char tx_data[TEST_BUF_SIZE] =
+	"It is harder to be kind than to be wise........";
+__aligned(CONFIG_DMA_BLEN_ALIGNMENT) char rx_data[TEST_BUF_SIZE] = {0};
diff --git a/tests/drivers/dma/chan_link_transfer/Kconfig b/tests/drivers/dma/chan_link_transfer/Kconfig
new file mode 100644
index 00000000000..3e0236541c0
--- /dev/null
+++ b/tests/drivers/dma/chan_link_transfer/Kconfig
@@ -0,0 +1,10 @@
+# Copyright (c) 2025 Tenstorrent AI ULC
+# SPDX-License-Identifier: Apache-2.0
+
+mainmenu "DMA Link Transfer Test"
+
+source "Kconfig.zephyr"
+
+config DMA_LINK_ALIGNMENT
+	int "TX and RX data alignment in bytes"
+	default 32
diff --git a/tests/drivers/dma/chan_link_transfer/src/test_dma.c b/tests/drivers/dma/chan_link_transfer/src/test_dma.c
index 78b52b77857..3bc8195fa1d 100644
--- a/tests/drivers/dma/chan_link_transfer/src/test_dma.c
+++ b/tests/drivers/dma/chan_link_transfer/src/test_dma.c
@@ -34,9 +34,10 @@ static __aligned(32) char rx_data[RX_BUFF_SIZE] __used
 static __aligned(32) char rx_data2[RX_BUFF_SIZE] __used
 	__attribute__((__section__(".nocache.dma")));
 #else
-static const char tx_data[] = "It is harder to be kind than to be wise........";
-static char rx_data[RX_BUFF_SIZE] = { 0 };
-static char rx_data2[RX_BUFF_SIZE] = { 0 };
+__aligned(CONFIG_DMA_LINK_ALIGNMENT) static const char tx_data[] =
+	"It is harder to be kind than to be wise........";
+__aligned(CONFIG_DMA_LINK_ALIGNMENT) static char rx_data[RX_BUFF_SIZE] = { 0 };
+__aligned(CONFIG_DMA_LINK_ALIGNMENT) static char rx_data2[RX_BUFF_SIZE] = { 0 };
 #endif
 
 static void test_done(const struct device *dma_dev, void *arg, uint32_t id,
@@ -53,7 +54,7 @@ static int test_task(int minor, int major)
 {
 	struct dma_config dma_cfg = { 0 };
 	struct dma_block_config dma_block_cfg = { 0 };
-	const struct device *const dma = DEVICE_DT_GET(DT_NODELABEL(dma0));
+	const struct device *const dma = DEVICE_DT_GET(DT_NODELABEL(tst_dma0));
 
 	if (!device_is_ready(dma)) {
 		TC_PRINT("dma controller device is not ready\n");
@@ -126,6 +127,7 @@ static int test_task(int minor, int major)
 		return TC_FAIL;
 	}
 	k_sleep(K_MSEC(2000));
+
 	TC_PRINT("%s\n", rx_data);
 	TC_PRINT("%s\n", rx_data2);
 	if (minor == 0 && major == 1) {
diff --git a/tests/drivers/dma/cyclic/Kconfig b/tests/drivers/dma/cyclic/Kconfig
index d53b052542a..7e067e3e0e8 100644
--- a/tests/drivers/dma/cyclic/Kconfig
+++ b/tests/drivers/dma/cyclic/Kconfig
@@ -12,3 +12,7 @@ config DMA_CYCLIC_CHANNEL_NR
 config DMA_CYCLIC_XFER_SIZE
 	int "Number of bytes to transfer"
 	default 4096
+
+config DMA_CYCLIC_ALIGNMENT
+	int "TX and RX data alignment in bytes"
+	default 32
diff --git a/tests/drivers/dma/cyclic/src/test_dma_cyclic.c b/tests/drivers/dma/cyclic/src/test_dma_cyclic.c
index c80d0e6d9d6..d2a0531fdf8 100644
--- a/tests/drivers/dma/cyclic/src/test_dma_cyclic.c
+++ b/tests/drivers/dma/cyclic/src/test_dma_cyclic.c
@@ -22,8 +22,8 @@
 #include <zephyr/drivers/dma.h>
 #include <zephyr/ztest.h>
 
-static __aligned(32) uint8_t tx_data[CONFIG_DMA_CYCLIC_XFER_SIZE];
-static __aligned(32) uint8_t rx_data[CONFIG_DMA_CYCLIC_XFER_SIZE];
+static __aligned(CONFIG_DMA_CYCLIC_ALIGNMENT) uint8_t tx_data[CONFIG_DMA_CYCLIC_XFER_SIZE];
+static __aligned(CONFIG_DMA_CYCLIC_ALIGNMENT) uint8_t rx_data[CONFIG_DMA_CYCLIC_XFER_SIZE];
 
 K_SEM_DEFINE(xfer_sem, 0, 1);
 
@@ -48,7 +48,7 @@ static int test_cyclic(void)
 		tx_data[i] = i;
 	}
 
-	dma = DEVICE_DT_GET(DT_ALIAS(dma0));
+	dma = DEVICE_DT_GET(DT_NODELABEL(tst_dma0));
 	if (!device_is_ready(dma)) {
 		TC_PRINT("dma controller device is not ready\n");
 		return TC_FAIL;
diff --git a/tests/drivers/dma/loop_transfer/Kconfig b/tests/drivers/dma/loop_transfer/Kconfig
index e06d84296b2..22e973a2cce 100644
--- a/tests/drivers/dma/loop_transfer/Kconfig
+++ b/tests/drivers/dma/loop_transfer/Kconfig
@@ -21,3 +21,7 @@ config DMA_LOOP_TRANSFER_NUMBER_OF_DMAS
 config DMA_LOOP_TRANSFER_SIZE
 	int "Number of bytes to transfer"
 	default 8192
+
+config DMA_LOOP_ALIGNMENT
+	int "TX and RX data alignment in bytes"
+	default 32
diff --git a/tests/drivers/dma/loop_transfer/src/test_dma_loop.c b/tests/drivers/dma/loop_transfer/src/test_dma_loop.c
index dd8a8633dde..126ab01472b 100644
--- a/tests/drivers/dma/loop_transfer/src/test_dma_loop.c
+++ b/tests/drivers/dma/loop_transfer/src/test_dma_loop.c
@@ -33,8 +33,9 @@
 
 #define TRANSFER_LOOPS (4)
 
-static __aligned(32) uint8_t tx_data[CONFIG_DMA_LOOP_TRANSFER_SIZE];
-static __aligned(32) uint8_t rx_data[TRANSFER_LOOPS][CONFIG_DMA_LOOP_TRANSFER_SIZE] = { { 0 } };
+static __aligned(CONFIG_DMA_LOOP_ALIGNMENT) uint8_t tx_data[CONFIG_DMA_LOOP_TRANSFER_SIZE];
+static __aligned(CONFIG_DMA_LOOP_ALIGNMENT) uint8_t
+	rx_data[TRANSFER_LOOPS][CONFIG_DMA_LOOP_TRANSFER_SIZE] = { { 0 } };
 
 volatile uint32_t transfer_count;
 volatile uint32_t done;
diff --git a/tests/drivers/dma/scatter_gather/Kconfig b/tests/drivers/dma/scatter_gather/Kconfig
index 57b855ba07a..45f95538214 100644
--- a/tests/drivers/dma/scatter_gather/Kconfig
+++ b/tests/drivers/dma/scatter_gather/Kconfig
@@ -12,3 +12,7 @@ config DMA_SG_CHANNEL_NR
 config DMA_SG_XFER_SIZE
 	int "Number of bytes to transfer"
 	default 8192
+
+config DMA_SG_ALIGNMENT
+	int "TX and RX data alignment in bytes"
+	default 32
diff --git a/tests/drivers/dma/scatter_gather/src/test_dma_sg.c b/tests/drivers/dma/scatter_gather/src/test_dma_sg.c
index c24bb689869..ebb526d7b47 100644
--- a/tests/drivers/dma/scatter_gather/src/test_dma_sg.c
+++ b/tests/drivers/dma/scatter_gather/src/test_dma_sg.c
@@ -30,8 +30,8 @@ static __aligned(32) uint8_t rx_data[XFERS][CONFIG_DMA_SG_XFER_SIZE] __used
 	__attribute__((__section__(".nocache.dma")));
 #else
 /* this src memory shall be in RAM to support using as a DMA source pointer.*/
-static __aligned(32) uint8_t tx_data[CONFIG_DMA_SG_XFER_SIZE];
-static __aligned(32) uint8_t rx_data[XFERS][CONFIG_DMA_SG_XFER_SIZE] = { { 0 } };
+static __aligned(CONFIG_DMA_SG_ALIGNMENT) uint8_t tx_data[CONFIG_DMA_SG_XFER_SIZE];
+static __aligned(CONFIG_DMA_SG_ALIGNMENT) uint8_t rx_data[XFERS][CONFIG_DMA_SG_XFER_SIZE] = {{0}};
 #endif
 
 K_SEM_DEFINE(xfer_sem, 0, 1);
@@ -66,7 +66,7 @@ static int test_sg(void)
 
 	memset(rx_data, 0, sizeof(rx_data));
 
-	dma = DEVICE_DT_GET(DT_ALIAS(dma0));
+	dma = DEVICE_DT_GET(DT_NODELABEL(tst_dma0));
 	if (!device_is_ready(dma)) {
 		TC_PRINT("dma controller device is not ready\n");
 		return TC_FAIL;
